x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportUpgradeJobModelSnapshotAction.java

```diff
@@ -0,0 +1,258 @@
+    private static final byte UPGRADE_FROM_MAJOR = Version.CURRENT.major;
+
+    private static final Logger logger = LogManager.getLogger(TransportUpgradeJobModelSnapshotAction.class);
+
+    private final XPackLicenseState licenseState;
+    private final PersistentTasksService persistentTasksService;
+    private final JobConfigProvider jobConfigProvider;
+    private final JobResultsProvider jobResultsProvider;
+    private final MlMemoryTracker memoryTracker;
+    private final MlConfigMigrationEligibilityCheck migrationEligibilityCheck;
+
+    @Inject
+    public TransportUpgradeJobModelSnapshotAction(Settings settings, TransportService transportService, ThreadPool threadPool,
+                                                  XPackLicenseState licenseState, ClusterService clusterService,
+                                                  PersistentTasksService persistentTasksService, ActionFilters actionFilters,
+                                                  IndexNameExpressionResolver indexNameExpressionResolver,
+                                                  JobConfigProvider jobConfigProvider, MlMemoryTracker memoryTracker,
+                                                  JobResultsProvider jobResultsProvider) {
+        super(UpgradeJobModelSnapshotAction.NAME, transportService, clusterService, threadPool, actionFilters, Request::new,
+            indexNameExpressionResolver, Response::new, ThreadPool.Names.SAME);
+        this.licenseState = licenseState;
+        this.persistentTasksService = persistentTasksService;
+        this.jobConfigProvider = jobConfigProvider;
+        this.jobResultsProvider = jobResultsProvider;
+        this.memoryTracker = memoryTracker;
+        this.migrationEligibilityCheck = new MlConfigMigrationEligibilityCheck(settings, clusterService);
+    }
+
+    @Override
+    protected ClusterBlockException checkBlock(Request request, ClusterState state) {
+        return state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_WRITE);
+    }
+
+    @Override
+    protected void masterOperation(Task task, Request request, ClusterState state,
+                                   ActionListener<Response> listener) {
+        if (migrationEligibilityCheck.jobIsEligibleForMigration(request.getJobId(), state)) {
+            listener.onFailure(ExceptionsHelper.configHasNotBeenMigrated("upgrade job snapshot", request.getJobId()));
+            return;
+        }
+
+        if (licenseState.checkFeature(XPackLicenseState.Feature.MACHINE_LEARNING) == false) {
+            listener.onFailure(LicenseUtils.newComplianceException(XPackField.MACHINE_LEARNING));
+            return;
+        }
+
+        if (state.nodes().getMaxNodeVersion().after(state.nodes().getMinNodeVersion())) {
+            listener.onFailure(ExceptionsHelper.conflictStatusException(
+                "Cannot upgrade job [{}] snapshot [{}] as not all nodes are on version {}. All nodes must be the same version",
```

Comment:
This just eliminates the small edge cases of requiring job node assignment to take into account node version. These processes are short lived, and restricting the cluster to not be a mixed cluster is a sane limitation. Especially since this API is meant to be used right before upgrading to the next major version.